gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+ 
+- Restart CoreDNS pods through the API server if pods are up and running but DNS reply does not match the expectation. 
+
+
+ 
+### Strategies for Different Deployment Scenarios 
+
+I have listed out four possible scenarios that the binary might be deployed in. Out of these, three have a determinate solution and can be implemented right away. The last one will require some hacking and also relies on a feature which is under development. 
+- **Binary Deployed as a Pod** 
+
+
+  Here, the team decides to launch the binary as a health check pod using `kubectl` and a YAML configuration. The steps listed above can then be directly executed with no additional configuration or modifications required.  
+
+
+- **Binary Deployed Externally but Creating/Deleting Pods Inside the Cluster Allowed** 
+![CoreDNSPod](https://user-images.githubusercontent.com/37150991/75658062-9220f900-5c8d-11ea-8c13-f48fbcecc37a.jpeg)
+ 
+ 
+  Here, the binary itself lives as an external program but will dynamically create a lightweight pod in the cluster (through communication with the API server). The only purpose of this pod inside the cluster would be to execute the dig command (and maybe other debug commands) that the external Golang application sends to this pod. 
+  
+  We can also get rid of the pod right after a check and create one the next time we are running a check so the main application doesn't actually live on the cluster and isn't affected by it. 
+
+
+  The pod is just created so that accessing the private k8s cluster network becomes simple (no routing or proxy required). It is straightforward to run a shell in a container (through kubectl exec) and then equally simple to run the dig commands in the shell. 
+
+
+- **Binary Deployed Externally and DNS Port Exposed** 
+![CoreDNSNodePort](https://user-images.githubusercontent.com/37150991/75658133-b5e43f00-5c8d-11ea-87de-2dc14f965cd3.jpeg)
+
+  Here, the DNS port or the UDP port 53 (in standard deployment) of the CoreDNS deployment is exposed through a service to the external world. The binary, deployed externally, can then use this service to ping the endpoints using the dig command.  
+ 
+
+- **Creation of Pods Forbidden and DNS Port Not Exposed**
+
+
+  In such a case, the binary (living as an external program) would have very limited options to interact with the cluster. However, there is a way this scenario can be tackled elegantly and that is through a port-forward.  
+
+
+  Port-forward creates a connection between a port on our application host and the port we want to connect to on a pod, inside the cluster.  Although port-forward works only for TCP connections currently, there is an active issue to bring the functionality to UDP ports and I’ll be tracking the issue over the next months to see if something can be hacked together.  
+
+
```

Comment:
@WJayesh , the thing you don't have here is the capturing of metrics associated with DNS resolution. Do you think it would be helpful to know if it took too long to get a DNS response?

Reply:
It is true that the metrics would help determine the state in a general case but this case talks about a situation where the DNS port is not accessible to the binary. Thus, without using port-forward, there can be no communication, no DNS response and therefore any metrics might not be useful in this case at all.

Reply:
> Thus, without using port-forward, there can be no communication, no DNS response and therefore any metrics might not be useful in this case at all.

@WJayesh, don't assume that. The user of the binary could have access to the overlay network and hence could have access to the pod directly.

Reply:
@greenpau 
>The user of the binary could have access to the overlay network and hence could have access to the pod directly

I understand this idea on a high level and will include this in the RFC. 
I'll, however, admit that I lack some clarity on how exactly I'll make the overlay network accessible to the health check binary. I am reading about it and will add new information if required to this point.

Reply:
Also, I learnt about your work in NYC and want to thank you for your service! Please take care of your health :)